Do not update endpoint state via gossiper::add_saved_endpoint once it was updated via gossip

bhalevy opened the pull request:
Currently, `add_saved_endpoint` is called from two paths:  One, is when loading states from system.peers in the join path (join_cluster, join_token_ring), when `_raft_topology_change_enabled` is false, and the other is from `storage_service::topology_state_load` when raft topology changes are enabled.

In the later path, from `topology_state_load`, `add_saved_endpoint` is called only if the endpoint_state does not exist yet.  However, this is checked without acquiring the endpoint_lock and so it races with the gossiper, and once `add_saved_endpoint` acquires the lock, the endpoint state may already be populated.

Since `add_saved_endpoint` applies local information about the endpoint state (e.g. tokens, dc, rack), it uses the local heart_beat_version, with generation=0 to update the endpoint states, and that is incompatible with changes applies via gossip that will carry the endpoint's generation and version, determining the state's update order.

This change makes sure that the endpoint state is never update in `add_saved_endpoint` if it has non-zero generation.
An internal error exception is thrown if non-zero generation is found, and in the only call site that might reach that state, in `storage_service::topology_state_load`, the caller acquires the endpoint_lock for checking for the existence of the endpoint_state, calling `add_saved_endpoint` under the lock only if the endpoint_state does not exist.

Fixes #16429

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5277/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6682/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5277/)
#### Build Details:
- Duration: 2 hr 3 min
- Builder: spider5.cloudius-systems.com

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -2088,7 +2088,7 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
         ep_state = *es;
         logger.debug("not replacing a previous ep_state for {}, but reusing it: {}", ep, ep_state);
-        ep_state.set_heart_beat_state_and_update_timestamp(heart_beat_state());
+        ep_state.update_timestamp();
```
We cannot reuse existing generation if we set endpoint state version using our local version generator, e.g. here:
```
 ep_state.add_application_state(gms::application_state::HOST_ID, versioned_value::host_id(host_id.value()));
```

Local node's highest version may be ahead of the ep_state node's highest version. If we gossip this state around, other nodes will ignore changes to states made by the original node. That's because gossip protocol is incremental, node's request only states with versions higher than the highest known version.

Setting generation to 0 works because it will always lose with the states gossiped by the original node.

bhalevy replied:
I've been thinking about that too actually.
Once the endpoint state is set with a non-zero generation we can assume it was set by gossip directly or indirectly from the node itself, since other nodes will add their saved info about it with generation=0, as we do (now), and so add_saved_endpoint should not update it in this case.
I don't think that was ever intended so it might even warrant an internal error.

bhalevy commented:
In v2 ([a94668d](https://github.com/scylladb/scylladb/commit/a94668d59d3bbb6b387471c4cc82720af742866d)):
- added `on_internal_error_noexcept` if local endpoint state found and has non-zero generation
  - just returns immediately in production
  - and added comment explaining why using local versioning for `versioned_value` is okay.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5303/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6707/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5303/)
#### Build Details:
- Duration: 2 hr 15 min
- Builder: spider6.cloudius-systems.com

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5302/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6711/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5302/)
#### Build Details:
- Duration: 4 hr 11 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented:
Dequeued, broke next promotions

https://jenkins.scylladb.com/job/scylla-master/job/next/6963/artifact/testlog/aarch64/debug/topology.test_topology_rejoin.1.log
https://jenkins.scylladb.com/job/scylla-master/job/next/6963/artifact/testlog/aarch64/debug/scylla-151.log

```
ERROR 2023-12-19 08:27:46,944 [shard 0:main] gossip - Attempted to add saved endpoint after endpoint_state was already established with gossip: { generation = 1702974301, version = 230 }, at: 0xcfc2a63 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x504b927 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x504b68f /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x504ddff /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x504e82b /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3a96b3f /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3a97333 0x1a7515e7 0xd4c8db7 0xd4c8987 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d48573 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d51327 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d54f13 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d5300f /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x379b4d3 /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3798bfb 0xd06425f 0xe0d3fe7 0xe0d3bab 0xe0d33bb 0xd293b53 0xd061c27 /lib64/libc.so.6+0x3059b /lib64/libc.so.6+0x3066f 0xcf7ca2f
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
   --------
   seastar::internal::coroutine_traits_base<void>::promise_type
Aborting on shard 0.
Backtrace:
  0xcfc2a63
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3f44adf
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3f446bb
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d266b7
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d569db
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3e8a47f
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3e8a7c7
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3e8a5eb
  linux-vdso.so.1+0x7db
  /lib64/libc.so.6+0x90a6f
  /lib64/libc.so.6+0x44b3f
  /lib64/libc.so.6+0x30287
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3a97353
  0x1a7515e7
  0xd4c8db7
  0xd4c8987
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d48573
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d51327
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d54f13
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3d5300f
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x379b4d3
  /jenkins/workspace/scylla-master/next/scylla/build/debug/seastar/libseastar.so+0x3798bfb
  0xd06425f
  0xe0d3fe7
  0xe0d3bab
  0xe0d33bb
  0xd293b53
  0xd061c27
  /lib64/libc.so.6+0x3059b
  /lib64/libc.so.6+0x3066f
  0xcf7ca2f
------ Ending test debug/topology.test_topology_rejoin.1::test_start_after_sudden_stop ------
```

`topology_state_load` is calling `add_saved_endpoints` after gossiping started.

bhalevy commented:
In v3 ([e0c2548](https://github.com/scylladb/scylladb/commit/e0c2548097a6ec54f6699511c7b42f41af0d1da3)):
- reduced on_internal_error_noexcept to warning (including a backtrace fwiw)
- storage_service: topology_state_load: lock endpoint for add_saved_endpoint
  - to prevent the warning (what caused the internal error above in aarch64/debug)

tgrabiec commented on gms/gossiper.cc:
```diff
@@ -2086,10 +2086,18 @@ future<> gossiper::add_saved_endpoint(inet_address ep) {
     auto es = get_endpoint_state_ptr(ep);
     if (es) {
+        if (es->get_heart_beat_state().get_generation()) {
+            logger.warn("Attempted to add saved endpoint after endpoint_state was already established with gossip: {}, at {}", es->get_heart_beat_state(), current_backtrace());
```
May be worthwhile to log "ep" so that we know for which endpoint we have a collision.

bhalevy replied:
Oops, definitely worth printing it.

gusev-p commented on service/storage_service.cc:
```diff
@@ -557,9 +557,13 @@ future<> storage_service::topology_state_load() {
         const auto ep = tmptr->get_endpoint_for_host_id(e);
-        if (!is_me(e) && !_gossiper.get_endpoint_state_ptr(ep)) {
-            co_await _gossiper.add_saved_endpoint(ep);
+        auto permit = co_await _gossiper.lock_endpoint(ep, gms::null_permit_id);
```
I don't understand something about `add_saved_endpoint` call in `storage_service::join_cluster` - we are load ing IPs/tokens from `system.peers` and constructing `token_metadata` from them. We call `add_saved_endpoint` in the loop, by it uses the old version of `token_metadata`, not the one we'are constructing. This means we probably never go under the `if (host_id) {` in it and never fill `HOST_ID` and `TOKENS`.

gusev-p replied:
there is another place where `add_saved_endpoint` is called - `storage_service::join_token_ring`:
```
co_await _gossiper.reset_endpoint_state_map();
for (auto ep : loaded_endpoints) {
    co_await _gossiper.add_saved_endpoint(ep);
}
```
so it probably works my this means. By why then we need the `add_saved_endpoint` call in `join_cluster`?

bhalevy replied:
In join_cluster we get the host_id from system.peers.
This is more evident in https://github.com/scylladb/scylladb/pull/15788/commits/a4b158d1d5d055d198514bb1af4fb1c38c6a9952

In `join_token_ring`, we call `add_saved_endpoints` after `do_shadow_round` which clears the endpoint states.
This is orthogonal to this PR, but I think it'd be possible to keep the loaded state in the gossiper (using `loaded_endpoint_state` to be introduced in `https://github.com/scylladb/scylladb/pull/15788`, and loaded automatically in `gossiper::reset_endpoint_state_map` so the reset endpoint state map will contain the state loaded from system.peers.

gusev-p replied:
> In join_cluster we get the host_id from system.peers.
> This is more evident in [a4b158d](https://github.com/scylladb/scylladb/commit/a4b158d1d5d055d198514bb1af4fb1c38c6a9952)

This is not "more evident", this is actually changed behaviour — IICU, without this commit we don't see `host_id` and `tokens` for ep in `join_cluster/add_saved_endpoint` call.

> This is orthogonal to this PR, but I think it'd be possible to keep the loaded state in the gossiper (using `loaded_endpoint_state` to be introduced in `https://github.com/scylladb/scylladb/pull/15788`, and apply the loaded endpoint state automatically in `gossiper::reset_endpoint_state_map` so the reset endpoint state map will contain the state loaded from system.peers.

This sounds cleaner.

kostja replied:
@bhalevy why do we need to call add_saved_endpoint more than once? (join_cluster, join_token_ring). It seems this needs to be cleared first.

bhalevy replied:
I explained it in some other comment in this PR IIRC.
The other call is needed after shadow round that resets the endpoint states map.
This is not needed to be cleared first because this all happens before gossip starts so the endpoint state has generation=0 anyhow so updating it more than once in this state via add_saved_endpoint is not a problem.

kostja replied:
So why does the shadow round reset the endpoint states map?

bhalevy replied:
I discussed the resetting of the endpoint state with @asias and maybe he can explain it better than me.

In any case, I'd like to avoid too intrusive changes in the scope of this PR to make it feasible for backport.
Let's consider design changes for improvement as follow up.

https://github.com/scylladb/scylladb/pull/16432#issuecomment-1869523583 is one such direction.

As for loading the state once, see https://github.com/scylladb/scylladb/pull/16432#discussion_r1431495073 earlier in this thread.
It'd be possible with https://github.com/scylladb/scylladb/pull/16432#issuecomment-1869523583, when `topology_state_load` will stop abusing `add_saved_endpoint` to inject application states, but rather it would that using an official api similar to `add_local_application_state`.

bhalevy replied:
But please, let's keep in mind that this is a point fix that should be backportable and hold our breath with fixing the whole world around it.

kostja replied:
So the patch has no test, the impact on the clients (is anyone suffering from this?) is unclear,  and is working around a problem that only @asias knows why it exists :(

@gusev-p do you understand why we have to add endpoints twice? @tgrabiec perhaps you?

kostja replied:
@bhalevy why should it be backportable?

bhalevy replied:
> @bhalevy why should it be backportable?

The context is https://github.com/scylladb/scylla-enterprise/issues/3518 and https://github.com/scylladb/scylladb/pull/15788

> So the patch has no test,

It is covered by the existing topology tests.

> the impact on the clients (is anyone suffering from this?) is unclear, and is working around a problem that only @asias knows why it exists :(

The was stumbled upon in review.  See https://github.com/scylladb/scylladb/pull/15788#discussion_r1428382450

You keep returning to "why we have to add endpoints twice?", I presume in `join_cluster` and `join_token_ring`, but this is really not problematic WRT to the endpoint state generation, since both calls happen before gossip starts gossiping - i.e. the endpoint state will have generation=0 in both cases.

The problem happens via `topology_state_load` that uses `add_saved_endpoint` to inject the node's state, yet it may lose the race to gossiper and discover in `add_saved_endpoint` that the state was already loaded, after the latter acquires the endpoint_lock.

To prevent that (and yes, we don't want to add the state more than once), we want `topology_state_load` to acquire the endpoint lock for checking if the endpoint_state already exists.
And to do that, we want https://github.com/scylladb/scylladb/pull/16507 to release the lock around waiting in error injection so that `topology_state_load` can grab it.

I hope that clarifies the background for this PR and #16507.

kostja replied:
1) I believe your response is contradictory, sorry. If the behavior is covered by the existing tests, then how is 3518 possibl?
2) I checked 3518 and the customer did not follow our topology operations best practices. The behavior is known and is addressed by the topology changes on Raft. Moreover, it's part of the test plan for this.

So what's the point of fixing this in gossip mode in 6.0? If you wish to satisfy the customer, perhaps direct these patches straight to the enterprise release branch.

bhalevy replied:
> 1. I believe your response is contradictory, sorry. If the behavior is covered by the existing tests, then how is 3518 possible?

I may have misunderstood your question then.
I'm talking about this PR and the access to the endpoint state via add_saved_endpoint that's covered, not about scylladb/scylla-enterprise#3518.
For the latter, I added dtest reproducers in https://github.com/scylladb/scylla-dtest/pull/3685
that naturally were added after it was hit.

> 2. I checked 3518 and the customer did not follow our topology operations best practices. The behavior is known and is addressed by the topology changes on Raft. Moreover, it's part of the test plan for this feature.

The fact that it's addressed by raft topology changes doesn't mean we don't want to fix that in older releases.
It will take considerable time until consistent topology changes reach the field in production, and FE (and me) want some solution in place until then.

> So what's the point of fixing this in gossip mode in 6.0? If you wish to satisfy the customer, perhaps direct these patches straight to the enterprise release branch.

No, this is not enterprise only code. The fixes must go through master.

And it's beyond satisfying the customer.
We have a hole in add_saved_endpoint that doesn't populate DC and RACK and the changes to support it are relatively simple.
Trying to over engineer it is like killing a spider with a sledgehammer :)

gusev-p commented on gms/gossiper.cc:
```diff
@@ -2074,22 +2074,30 @@ void gossiper::build_seeds_list() {
     auto es = get_endpoint_state_ptr(ep);
     if (es) {
+        if (es->get_heart_beat_state().get_generation()) {
+            logger.warn("Attempted to add saved endpoint after endpoint_state was already established with gossip: {}, at {}", es->get_heart_beat_state(), current_backtrace());
```
> reduced on_internal_error_noexcept to warning (including a backtrace fwiw)

Why? The problem with `add_saved_endpoint` call from `topology_state_load` was fixed mainly because it complained loudly. Nobody will notice this `warn`-s.

bhalevy replied:
With `co_return` after this warning the issue is benign.
I have second thoughts about the internal error.
I think that the right direction would be to automatically restore the loaded state in `gossiper::reset_endpoint_state_map` as I described above, and then `gossiper::add_saved_endpoint` would either go away or become private in class gossiper.

gusev-p replied:
> With `co_return` after this warning the issue is benign.

But it still breaks the assumptions of `add_saved_endpoint — it shouldn't be called for existing ep, at least this is what is maintained in the current PR. Broken invariants should cause `inernal_error`.

I added `add_saved_endpoint` into `topology_state_load` some time ago, but now I don't like it. It incurs an overhead of grabbing the gossiper lock for every ep in the token_metadata each time `topology_state_load` is called, but `add_saved_endpoint` is only needed on node startup IIUC. Maybe we can do `add_saved_endpoint` in `topology_state_load` only once, on the first call.

tgrabiec replied:
Don't we also need add_saved_endpoint() later than boot, if the node missed some topology changes and then catches up from group0 leader?

bhalevy replied:
This is discussed below...

kbr-scylla replied:
@tgrabiec please also take a look at the proposal here https://github.com/scylladb/scylladb/issues/16644

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5319/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6732/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5319/)
#### Failed Tests (9/23473):
- [test_disablebinary_and_disablegossip](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/nodetool_additional_test/TestNodetool/Tests___dtest___test_disablebinary_and_disablegossip) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_disablebinary_and_disablegossip)
- [test_old_ip_notification_repro](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/test_old_ip_notification_repro/Tests___Unit_Tests___test_old_ip_notification_repro) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_old_ip_notification_repro)
- [test_old_ip_notification_repro](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/test_old_ip_notification_repro/Tests___Unit_Tests___test_old_ip_notification_repro_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_old_ip_notification_repro)
- [test_old_ip_notification_repro](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/test_old_ip_notification_repro/Tests___Unit_Tests___test_old_ip_notification_repro_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_old_ip_notification_repro)
- [test_raft_replace_ignore_nodes](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/test_raft_ignore_nodes/Tests___Unit_Tests___test_raft_replace_ignore_nodes) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_raft_replace_ignore_nodes)
- [topology_custom.test_old_ip_notification_repro.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_old_ip_notification_repro_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_old_ip_notification_repro.dev.1)
- [topology_custom.test_old_ip_notification_repro.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_old_ip_notification_repro_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_old_ip_notification_repro.dev.2)
- [topology_custom.test_old_ip_notification_repro.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_custom_test_old_ip_notification_repro_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_custom.test_old_ip_notification_repro.dev.3)
- [topology_experimental_raft.test_raft_ignore_nodes.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/5319/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_raft_ignore_nodes_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_raft_ignore_nodes.dev.2)
#### Build Details:
- Duration: 4 hr 4 min
- Builder: i-0940bef5728bae2e7 (m5d.8xlarge)

bhalevy commented:
- Rebased on top of https://github.com/scylladb/scylladb/pull/16507 (and made draft)
- Reversed patches to support locking be caller around add_saved_endpoint first
  - and call it from `topology_state_load` only if the endpoint state doesn't exist
- Then, apply "gossiper: add_saved_endpoint: keep heart_beat_state if ep_state is found"
  - and call on_internal_error if an endpoint_state for this endpoint exists and has non-zero generation, as now all call sites must conform to that.

bhalevy commented:
In the future, I suggest to consider refactoring the endpoint_state_map out of the gossiper so it can be updated both from gossip and from raft.

With `_raft_topology_change_enabled`, raft could "own" the topology related states and gossip can deal with the other states.  To allow such a hybrid model, I think that the heartbeat_state (generation + version) should be moved from the endpoint state level to the application state level, so some application states can be safely managed from raft, using the local generation/version, while the other app states can be managed using generation 0 if and when they are loaded from system.peers and then using the remote generation/version when gossip gets the app states from remote peers and from the endpoint itself.  In the latter case, the non-zero generation always originates from the endpoint itself, just that some peers may still have an older generation from a previous incarnation of the node, in case it restarted.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5443/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6857/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5443/)
#### Build Details:
- Duration: 1 hr 30 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
In v4 ([88eb36e](https://github.com/scylladb/scylladb/commit/88eb36e4c6182fc01969447beeded390b371a9aa)):
- raft_group_registry: move on_alive error injection to gossiper
  - Replaces https://github.com/scylladb/scylladb/pull/16507

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5474/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6893/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5474/)
#### Build Details:
- Duration: 1 hr 44 min
- Builder: spider6.cloudius-systems.com

gusev-p commented on gms/gossiper.cc:
```diff
@@ -1616,6 +1617,24 @@ void gossiper::mark_alive(inet_address addr) {
 }
 
 future<> gossiper::real_mark_alive(inet_address addr) {
+    co_await utils::get_local_injector().inject_with_handler("raft_group_registry::on_alive", [this, endpoint = addr] (auto& handler) -> future<> {
```
it's better to rename the injection name, e.g. `gossiper::real_mark_alive`

bhalevy replied:
no problem

bhalevy replied:
done in v5

gusev-p commented on test/topology_custom/test_old_ip_notification_repro.py:
```diff
@@ -21,7 +21,7 @@ async def test_old_ip_notification_repro(manager: ManagerClient) -> None:
     """
     Regression test for #14257.
     It starts two nodes. It introduces a sleep in raft_group_registry::on_alive
```
this commit needs to be adjusted (`raft_group_registry::on_alive`)

bhalevy replied:
it was supposed to indicate the error injection point
but if we're changing it as per the above I'll change it here too

bhalevy replied:
done in v5

bhalevy commented:
In v5 (3cba079):
- renamed error injection point to gossiper::real_mark_alive and adjusted comment in test.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5477/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5477/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6897/)
#### Build Details:
- Duration: 1 hr 39 min
- Builder: spider6.cloudius-systems.com

kbr-scylla commented on gms/gossiper.cc:
```diff
@@ -2105,10 +2105,18 @@ future<> gossiper::add_saved_endpoint(inet_address ep, permit_id pid) {
     auto ep_state = endpoint_state();
     auto es = get_endpoint_state_ptr(ep);
     if (es) {
+        if (es->get_heart_beat_state().get_generation()) {
```
As I wrote in https://github.com/scylladb/scylladb/issues/16429

To me it seems that nothing has actually changed.

First, the commit message suggests that the existing code "updates its heart_beat_state".

But assuming the assertion you put here is indeed true, i.e. add_saved_endpoint is only called when generation == 0, then the existing code does nothing. Indeed, it sets heart_beat_state to `heat_beat_state()` i.e. generation = 0, version = 0. But if generation == 0, that should actually not change anything.

So the only thing this code seems to do, is to actually add the assertion, not change any behavior.

bhalevy replied:
There is a subtle change where `topology_state_load` acquires the endpoint lock for checking if the endpoint state already exists.  Previously, it did so without acquiring the lock and so it could race with `add_saved_endpoint` that could have found such a state after acquiring the lock, and that state could have non-zero generation.

That's the gist of the fix.

kbr-scylla replied:
Ok, that makes sense, but can you please update the PR cover letter and commit messages (last one at least) so they explain this? I.e. the fact that we actually never have to "keep existing heart_beat_state" (other than 0), there was the possibility that it was non-zero when we called add_save_endpoint, but that was a bug which this PR fixes (second commit, as I understand) and now adds an assertion to prevent the bug from reoccuring.

kbr-scylla replied:
Also...

> it could race with add_saved_endpoint

so both topology_state_load and **some other place** would concurrently call `add_saved_endpoint`
what is that other place?

And shouldn't we ensure that this other place is always called first (e.g. before we start Raft group 0 server), instead of introducing locks?
cc @gusev-p

bhalevy replied:
No, the race is with `do_apply_state_locally`.

kbr-scylla replied:
> AFAIU topology_state_load is called in normal operation mode, not only when raft group is starts.

Yes, it's called on every topology change, for example.

> No, the race is with do_apply_state_locally.

So wait. But that means `add_saved_endpoint`, the one that is being called from `topology_state_load`, **will** observe non-zero generations, no? Because gossiping is happening in the background and is updating endpoint states received from nodes and setting non-zero generations from them. And then when we perform a topology change, `topology_state_load` calls `add_saved_endpoint` and your assertion should get triggered, no?

Now I'm even more confused -- why the tests are passing in CI...

>  That's why I think calling add_saved_endpoint there is an abuse of the interface.

I agree and @gusev-p also mentioned in another comment he doesn't like it.

I think in raft-topology mode, `add_saved_endpoint` should only be called once for each known peer **before** we start group 0 server (and outside `topology_state_load`), and its only purpose would be to restore known IP mappings from before the node stopped. Once that is done, we should never be calling `add_saved_endpoint`.

gusev-p replied:
>  And then when we perform a topology change, topology_state_load calls add_saved_endpoint and your assertion should get triggered, no?

`add_saved_endpoint` is only called for non-existing IP-s, which should have zero generation/version. The race this PR fixes is that `do_apply_state_locally` might have been called after the if in `topology_state_load` which checks `endpoint_state` doesn't exist.

gusev-p replied:
> I think in raft-topology mode, add_saved_endpoint should only be called once for each known peer before we start group 0 server (and outside topology_state_load), and its only purpose would be to restore known IP mappings from before the node stopped. Once that is done, we should never be calling add_saved_endpoint.

The reason I added `add_saved_endpoint` into `topology_state_load` is that it's the place where the current raft topology become known. We can't use `system.topology` before group0 start because it may be stale.

kbr-scylla replied:
Ok, so how about my proposal -- calling it before we start group 0 server, based on contents of system.peers table? Exactly like we do for gossip-based topology changes? Nodes that are appearing during normal operation (not when we restart) are not "saved endpoints", to it seems weird to call add_saved_endpoint from inside topology_state_load which is used during normal operation. If I understand the purpose of add_saved_endpoint, it's for restoring gossiper state on restart that we had before restart.

bhalevy replied:
In the future, I think we should consider a refactoring of the endpoint states map and allow updating it both from gossip and from raft.  In that model we may want a heart_beat_version per application state rather than per endpoint state.

This was raft can manage the topology-related states while gossip can update auxiliary states like the heat map, large partitions, etc.

kbr-scylla replied:
> The reason I added add_saved_endpoint into topology_state_load is that it's the place where the current raft topology become known. We can't use system.topology before group0 start because it may be stale.

Ah, but gossiper state that is supposed to be restored by add_saved_endpoint is not persisted in system.topology, it's persisted in system.peers.

And we don't care about the fact that it's stale, we're restarting and it was stale the same way as it was before we restarted. Further gossiping will update it.

bhalevy replied:
Or we can rely on class topology to hold all the critical stuff and it will be the source of truth.

gusev-p replied:
> And we don't care about the fact that it's stale, we're restarting and it was stale the same way as it was before we restarted. Further gossiping will update it.

Is it guaranteed that all the endpoints we see in raft topology will somehow eventually appear in gossiper? I see that we `remove_endpoint`-s from gossiper in `topology_state_load`, but we don't add them except through `add_saved_endpoint`.

And why not use the `system.topology` instead of `system.peers` than? `system.peers` is considered legacy.

kbr-scylla replied:
> Is it guaranteed that all the endpoints we see in raft topology will somehow eventually appear in gossiper?

Nodes are gossiping on their own. If a node is in raft topology and is alive, then it's gossiping with other nodes in raft topology. So they should see each other.

> I see that we remove_endpoint-s from gossiper in topology_state_load, but we don't add them except through add_saved_endpoint.

We never carefully designed how the interaction between raft-topology and gossiper should look like... Perhaps we should do it now.

I think that the only things that raft-topology should do to gossiper are:
- to tell it, through host-ID based interface, which nodes are banned. Gossiper remove all information about and would not further interact with nodes whose host IDs are banned
- update certain application states sometimes (tokens, status). But this could be done only for local node, no need to do it for peers. Nodes would update peers' information in gossiper through regular gossiping.

> And why not use the system.topology instead of system.peers then? system.peers is considered legacy.

In raft-topology mode, system.peers still has a single role that is required (but not legacy) -- persisting IP mappings and restoring them on boot. And that is also the only thing that we would like to restore into gossiper. system.topology doesn't have mappings.

kbr-scylla replied:
> If a node is in raft topology and is alive, then it's gossiping with other nodes in raft topology.

To be precise.. it's gossiping with nodes that it can reach through transitive closure obtained from seeds graph. But those are the only nodes it can gossip with --- even if raft topology contains information about certain node that is not in this transitive closure, we won't be able to gossip with it, because the transitive closure is the only way we can obtain the node's IP, which is needed to gossip with it.

**Raft topology can only restrict the set of nodes we're gossiping with, not extend it**

bhalevy replied:
FYI, one more site that I'd like to (ab)use add_saved_endpoint to inject an endpoint state, as a followup, is in prepare_replacement_info for ignored dead nodes.  This is required since the replacing node as no system.peers yet and those dead nodes qre visbile after full-cluster restart only in the gossip shadow round because they are not actively gossiping.

See https://github.com/scylladb/scylladb/pull/15788

kbr-scylla replied:
I propose we move this discussion somewhere else.
I created a new issue where I propose a solution. https://github.com/scylladb/scylladb/issues/16644

bhalevy replied:
> I propose we move this discussion somewhere else. I created a new issue where I propose a solution. #16644

Thanks.  I suggest going with this change for the time being and then look for improvements as follow up.

kbr-scylla commented:
Qd
